analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/ObjectCreation.cs

```diff
@@ -49,4 +56,12 @@ protected override ProgramState Process(SymbolicContext context, IObjectCreation
             return context.SetOperationConstraint(ObjectConstraint.NotNull);
         }
     }
+
+    private static CollectionConstraint CollectionCreationConstraint(ProgramState state, IObjectCreationOperationWrapper operation) =>
+        operation.Arguments.SingleOrDefault(IsEnumerable) is { } argument
+            ? state.Constraint<CollectionConstraint>(argument)
+            : CollectionConstraint.Empty;
+
+    private static bool IsEnumerable(IOperation operation) =>
+            operation.ToArgument().Parameter.Type.DerivesOrImplements(KnownType.System_Collections_IEnumerable);
```

Comment:
I have never checked... is there a way to break this code with a dynamic invocation or is `.ToArgument()` safe here?

Reply:
If the argument is dynamic, the entire operation is typed as `DynamicObjectCreation` and this processor will not be used.
For example:

```csharp
class Thingy
{
    void Work(dynamic d)
    {
        var res1 = new Thingy(d);
        var res2 = new Thingy(d.SomeInt);
    }

    public Thingy(int x) { }
}
```

![image](https://github.com/SonarSource/sonar-dotnet/assets/115458417/bd039135-1e3f-43ee-a236-63e33f67ff97)

...and [we will use](https://github.com/SonarSource/sonar-dotnet/blob/f75c2b4a8032f378257d0e23da60792d913b7a0a/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationDispatcher.cs#L52) the `NotNull` processor.